File: Python3/most-stones-removed-with-same-row-or-column.py
class Solution:
    def removeStones(self, stones: List[List[int]]) -> int:
        def dfs(i, j):
            points.discard((i, j))
            for y in rows[i]:
                if (i, y) in points:
                    dfs(i, y)
            for x in cols[j]:
                if (x, j) in points:
                    dfs(x, j)
        
        points = {(i, j) for i, j in stones}
        islands = 0
        rows = collections.defaultdict(list)
        cols = collections.defaultdict(list)

        for i, j in stones:
            rows[i].append(j)
            cols[j].append(i)

        for i, j in stones:
            if (i, j) in points:
                dfs(i, j)
                islands += 1

        return len(stones) - islands

File: Go/most-stones-removed-with-same-row-or-column.go
func removeStones(stones [][]int) int {
	points := map[pair]struct{}{}
	rows := map[int][]int{}
	cols := map[int][]int{}

	for _, stone := range stones {
		i, j := stone[0], stone[1]
		points[pair{i, j}] = struct{}{}
		rows[i] = append(rows[i], j)
		cols[j] = append(cols[j], i)
	}

	var dfs func(i, j int)
	dfs = func(i, j int) {
		delete(points, pair{i, j})
		for _, y := range rows[i] {
			if _, ok := points[pair{i, y}]; ok {
				dfs(i, y)
			}
		}
		for _, x := range cols[j] {
			if _, ok := points[pair{x, j}]; ok {
				dfs(x, j)
			}
		}
	}

	islands := 0

	for _, stone := range stones {
		i, j := stone[0], stone[1]
		if _, ok := points[pair{i, j}]; ok {
			dfs(i, j)
			islands++
		}
	}

	return len(stones) - islands
}

type pair [2]int